roles/translator/src/main.rs

```diff
@@ -77,6 +81,129 @@ async fn main() {
         broadcast::Receiver<server_to_client::Notify>,
     ) = broadcast::channel(10);
 
+    let task_collector: Arc<Mutex<Vec<(AbortHandle, String)>>> = Arc::new(Mutex::new(Vec::new()));
+
+    start(
+        rx_sv2_submit_shares_ext.clone(),
+        tx_sv2_submit_shares_ext.clone(),
+        tx_sv2_new_ext_mining_job.clone(),
+        tx_sv2_set_new_prev_hash.clone(),
+        tx_sv2_extranonce.clone(),
+        rx_sv2_extranonce.clone(),
+        rx_sv2_set_new_prev_hash.clone(),
+        rx_sv2_new_ext_mining_job.clone(),
+        rx_sv1_downstream.clone(),
+        tx_sv1_bridge.clone(),
+        tx_sv1_notify.clone(),
+        target.clone(),
+        tx_status.clone(),
+        task_collector.clone(),
+        proxy_config.clone(),
+    )
+    .await;
+
+    debug!("Starting up signal listener");
+    let task_collector_ = task_collector.clone();
+
+    let mut interrupt_signal_future = Box::pin(tokio::signal::ctrl_c().fuse());
+    debug!("Starting up status listener");
+    // Check all tasks if is_finished() is true, if so exit
+    loop {
+        let task_status = select! {
+            task_status = rx_status.recv().fuse() => task_status,
+            interrupt_signal = interrupt_signal_future => {
+                match interrupt_signal {
+                    Ok(()) => {
+                        info!("Interrupt received");
+                    },
+                    Err(err) => {
+                        error!("Unable to listen for interrupt signal: {}", err);
+                        // we also shut down in case of error
+                    },
+                }
+                break;
+            }
+        };
+        let task_status: Status = task_status.unwrap();
+
+        match task_status.state {
+            // Should only be sent by the downstream listener
+            State::DownstreamShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+                break;
+            }
+            State::BridgeShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+                break;
+            }
+            State::UpstreamShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+
+                // wait a random amount of time between 0 and 3000ms
```

Comment:
Why do you need to introduce rand as a dep for this? just go for the 3000ms ?

Reply:
as the comment states:

> if all the downstreams try to reconnect at the same time, the upstream may fail

if I understand correctly, this is trying to avoid a situation where multiple tProxies attempt to reconnect to the same Upstream at the same time, which could potentially cause it to fail

so it's attempting to introduce a randomness to this process, so that different tProxies will try to reconnect at different moments

I agree with the overall sentiment that this approach feels a bit hacky.

@lorbax do you know for a fact that the upstream will fail under these circumstances, or is it still just speculation?

if we're not sure this will happen, I'd say we should avoid solutions to imaginary problems

if we eventually see this happening, we can try to come up with a more sustainable solution

Reply:
this is best practice, when you have several client that can disconnect and then reconnect automatically you have to add a little randomness to the process. Is not hacky is the standard way to do this kind of thing

Reply:
what if an attacker deploys multiple malicious tProxies and intentionally resets all the connections at the same time?

Reply:
is there to prevent to ddos yourself or the pool that you are using

Reply:
@Fi3 could you please explain further? I dont understand why this is needed

Reply:
you have 1M tproxy connected to a pool. The pool have to close the connection, all the tproxy try to reconnect in the same exact moment. Not good

Reply:
I'm not strictly opposed to having this here, I just don't think there any substantial benefit

sure, we are making sure SRI tProxy is a "benign" client that intentionally avoids DDoSing the upstream connection

but in production scenarios, this is not really protecting upstream from malicious users

as discussed in the call, the only sustainable solution for this is load balancers and similar techniques to be deployed on the server infrastructure

Reply:
again this is industry best practice everyone does it. And there is a reason. This have nothing to do with protecting upstream from malicious users.  https://www.google.com/search?client=firefox-b-d&q=random+delay+backoff

Reply:
when you have an automatic reconnection or you retry to reconnect with exponetial backoff you add a random delay to avoid have all the client reconnect in the same instant.

Reply:
One thing is not ddos yourself another thing is protecting against a ddos

Reply:
@Fi3 Were this concerns addressed in https://github.com/stratum-mining/stratum/pull/1001#issuecomment-2270026282 it seems @lorbax agrees with the proposal, but it's not yet been committed?

Reply:
I think so. I did not reviewed it, I was only explaining why we should use randomized delay when we reconnect.

Reply:
@lorbax can you commit what @plebhash suggested if it addresses the concern? cc @plebhash 

Then we may be able to finally push this PR through and start testing it, and if it works https://github.com/stratum-mining/stratum/pull/1001#discussion_r1704210248 can be done in a follow up.